bullet-featherstone/src/JointFeatures.cc

```diff
@@ -0,0 +1,254 @@
+  }
+
+  if (std::get_if<RootJoint>(&joint->identifier))
+  {
+    const auto *model = this->ReferenceInterface<ModelInfo>(joint->model);
+
+    if (_dof < 3)
+      return model->body->getBaseVel()[_dof];
+    else if (_dof < 6)
+      return model->body->getBaseOmega()[_dof];
+  }
+
+  return gz::math::NAN_D;
+}
+
+/////////////////////////////////////////////////
+double JointFeatures::GetJointAcceleration(
+    const Identity &/*_id*/, const std::size_t /*_dof*/) const
+{
+  return gz::math::NAN_D;
+}
+
+/////////////////////////////////////////////////
+double JointFeatures::GetJointForce(
+    const Identity &_id, const std::size_t _dof) const
+{
+  const auto *joint = this->ReferenceInterface<JointInfo>(_id);
+  const auto *identifier = std::get_if<InternalJoint>(&joint->identifier);
+  if (identifier)
+  {
+    const auto *model = this->ReferenceInterface<ModelInfo>(joint->model);
+    return model->body->getJointTorqueMultiDof(
+          identifier->indexInBtModel)[_dof];
+  }
+
+  return gz::math::NAN_D;
+}
+
+/////////////////////////////////////////////////
+Pose3d JointFeatures::GetJointTransform(const Identity &_id) const
+{
+  (void) _id;
+  gzwarn << "Dummy function GetJointTransform\n";
+  return Pose3d();
+}
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointPosition(
+  const Identity &_id, const std::size_t _dof, const double _value)
+{
+  const auto *joint = this->ReferenceInterface<JointInfo>(_id);
+  const auto *identifier = std::get_if<InternalJoint>(&joint->identifier);
+  if (!identifier)
+    return;
+
+  const auto *model = this->ReferenceInterface<ModelInfo>(joint->model);
+  model->body->getJointPosMultiDof(identifier->indexInBtModel)[_dof] = _value;
+}
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointVelocity(
+  const Identity &_id, const std::size_t _dof, const double _value)
+{
+  const auto *joint = this->ReferenceInterface<JointInfo>(_id);
+  const auto *identifier = std::get_if<InternalJoint>(&joint->identifier);
+  if (!identifier)
+    return;
+
+  const auto *model = this->ReferenceInterface<ModelInfo>(joint->model);
+  model->body->getJointVelMultiDof(identifier->indexInBtModel)[_dof] = _value;
+}
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointAcceleration(
+  const Identity &/*_id*/, const std::size_t /*_dof*/, const double /*_value*/)
+{
+  // Do nothing
+}
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointForce(
+    const Identity &_id, const std::size_t _dof, const double _value)
+{
+  const auto *joint = this->ReferenceInterface<JointInfo>(_id);
+  const auto *identifier = std::get_if<InternalJoint>(&joint->identifier);
+  if (!identifier)
+    return;
+
+  const auto *model = this->ReferenceInterface<ModelInfo>(joint->model);
+  model->body->getJointVelMultiDof(identifier->indexInBtModel)[_dof] = _value;
```

Comment:
Thanks for catching that copy/paste error. It's funny that it accidentally gave the desired behavior, but only in Debug mode.

Making this change causes extreme instability, so ... that raises the question of whether there are other issues lurking around. I've tried playing with the PID gains in the NAO model and making dramatic reductions to P and I helped to legs to be more sensible, but the shoulders are still going wild.

In any case I've updated the PR with this fix. Now we can start debugging the instability instead of the nonsense behavior.